### ISSUE TITLE
Drag-and-drop support in 3D view

### ISSUE THREAD

**Issue by aymanhab** [opensim-org/opensim-gui#422]
Need to investigate what it takes to support dropping .osim, .mot files in embedded browser.

**Comment by jenhicks** [comment]
At least investigate implementing before the public beta.

**Comment by aymanhab** [comment]
This is unlikely to work as is because for security purposes, "browsers" do not have access o full file paths, sticky issue is file path needed to locate Geometry. Possible workarounds:
1. Drag and drop into navigator window (as a Java application doesn't have security issues)
2. Advertise or indicate the place where you can drag and drop (next to 3D window tab)
3. Make models into zip files that contain their own Geometry (long term)

**Comment by aseth1** [comment]
Opening anywhere in the Application window would be best. Worst case disable any drag/drop event on the 3D window itself.

**Comment by aseth1** [comment]
I don't believe this issue has been fixed.  Using this build [OpenSim-dd7dc47b-2018-02-11](https://ci.appveyor.com/project/opensim-org/opensim-gui/build/1.0.961/artifacts) when #514 was merged. Dragging the leg6dof9mus.osim model yields:
![image](https://user-images.githubusercontent.com/6494122/36459658-83b5547c-1669-11e8-8ecc-f88857cbc627.png)

Closing the Visualizer Window and dragging into the blank space does load the model. Closing the model and dragging into the Visualizer Window yields:
![image](https://user-images.githubusercontent.com/6494122/36459434-9cd7569a-1668-11e8-8695-12fe41415532.png)

Dragging into the **Models - properties** panel also loads the model but not the **Navigator**, which 
shows the *do not enter* sign.
![image](https://user-images.githubusercontent.com/6494122/36460090-b5b2674c-166b-11e8-91db-a04de67ec995.png)

I think we should be able to drag directly into the *Navigator* as well and get the  *do not enter* sign when dragging into the Visualizer Window (if it cannot be supported) rather than a cryptic message. It is definitely confusing to say we do not support .osim files!

Also the curso changes from Copy+ to Move->  and both of these seem incorrect. It should be "Open" or "Load".

**Comment by aymanhab** [comment]
Let me explain the change in behavior to see if we can find reasonable consensus:
- In 3.3 users had to drop .osim, .mot, .sto files into 3D window (drag handle displayed +Copy)
- In 4.0 users can drop files into toolbar, Properties Window, Coordinates Window (drag handle displayed +Move)
- Depending on whether graphics window is open, either you get the xml file to edit or a message that .osim files are not supported. 
I just clarified the message in the second case to state that OpenSim files need to be Dropped into Toolbar. I can block dropping altogether in the Browser window though that may prevent the possibility of dragging geometry into the scene and users may miss that they can drop into Toolbar or other windows. Let me know what you think @aseth1 @chrisdembia @tkuchida

**Comment by chrisdembia** [comment]
One solution is for the graphics window to be open when you first launch the GUI.

**Comment by aymanhab** [comment]
Thanks @chrisdembia but the reason we didn't do that (open gfx window on launch) in the first place is that the application GUI comes up before the server is ready to accept connections leading to Error/page-not-found misleading message in some cases.

**Comment by chrisdembia** [comment]
I vaguely remember that, but perhaps there is some way to get a callback when the GUI is done loading.

**Comment by aseth1** [comment]
@aymanhab I think the error window/message with the host address and "Unsupported file format (.osim)" should be removed. In its place we should block users from dragging into any region we do not want them too (or cannot support) by changing the cursor into a "do-not-enter" symbol and where it can be opened or loaded it should say "Load" or "Open" not Move or Copy since this is not the action being performed. I don't think the behavior of 3.3 is very relevant here especially if we are concerned with improving the interface. What we have now (OpenSim c596914-2018-02-24) isn't acceptable.

**Comment by aymanhab** [comment]
That's exactly what I implemented already, please verify when the build is
posted.

**Comment by aseth1** [comment]
Great. I verified that this is now the case on Windows with build OpenSim-2416f17a-2018-02-25.
Is there a reason we cannot drag onto the Navigator? That seems to be the most natural place to drag to.

**Comment by aymanhab** [comment]
Thanks for verifying. The reason dropping of files into trees and lists is
blocked by default is that there's built in support for drag and drop
within these widgets e.g. to reorder that we don't want to hijack and could
be useful to utilize in the future.

**Comment by jenhicks** [comment]
This does not work for me on my Mac. @aymanhab Is that expected?

**Comment by aymanhab** [comment]
Dropping into toolbar (and some other windows) should work, dropping into 3D view or navigator should not work for the reason mentioned above. It should be documented though since users may have been used to it from previous releases.

**Comment by jenhicks** [comment]
OK, the behavior works on Mac as you describe, although this does seem to be a downgrade from 3.3. The navigator and the 3D window are where I would try to drop a model. So I think most people will think the feature is no longer supported. Are there any other options?

**Comment by jenhicks** [comment]
That being said, I almost never use drag and drop on any application. 🦋  I think I may be the exception though.

**Comment by chrisdembia** [comment]
I usually don't use drag and drop but I'm learning that it is pretty handy. I think @aseth1 uses it extensively.

**Comment by aymanhab** [comment]
In tools e.g. VisualStudio, you can drop anywhere, but dropping in navigator/tree has specific meaning (e.g. add to project). Eventually we want to leverage that and allow drag and drop in navigator between models/submodels etc. If we allow drop in Navigator now we may open a big can of new issues.
